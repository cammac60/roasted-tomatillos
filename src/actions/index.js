export const addMovies = movies => ({
  type: 'ADD_MOVIES',
  movies
})

export const addLoaded = isLoaded => ({
  type: 'IS_LOADED',
  isLoaded
})

export const hasError = error => ({
  type: 'HAS_ERROR',
  error
})
