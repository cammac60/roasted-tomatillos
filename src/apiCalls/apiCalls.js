export const getMovies = async () => {
  const url = 'https://rancid-tomatillos.herokuapp.com/api/v1/movies'
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error('There was an error getting movies.');
  }
  const movies = await response.json();
  return movies
}
