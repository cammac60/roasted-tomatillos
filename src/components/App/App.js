import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import './App.scss';
import MoviesContainer from '../../containers/MoviesContainer/MoviesContainer';
import Header from '../Header/Header.js';
import { Route } from 'react-router-dom';
import { getMovies } from '../../apiCalls/apiCalls';
import { addMovies, addLoaded, hasError } from '../../actions';

export class App extends Component {
  constructor(props) {
    super(props)
  }
  
  componentDidMount() {
    this.fetchMoviesData()
  }

  fetchMoviesData = async () => {
    const { addMovies, addLoaded, hasError} = this.props
    try {
      const result = await getMovies()
      addMovies(result.movies)
      addLoaded(true)
    }
    catch (error) {
      addLoaded(false)
      hasError(error.message)
    }
  }

  render() {
    return (
      <div className="App">
        <Route exact path='/' render={() =>
          <>
            <Header />
            <MoviesContainer />
          </>
        }/>
        <Route path="/login" render={() => <div color="white">Temporary text</div>}/>
      </div>
    );
  }
}

export const mapDispatchToProps = dispatch => (
  bindActionCreators({
    addMovies,
    addLoaded,
    hasError
  }, dispatch)
)

export default connect(null, mapDispatchToProps)(App);
