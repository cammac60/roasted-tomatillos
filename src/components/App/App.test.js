import React from 'react';
import { shallow } from 'enzyme';
import { App, mapDispatchToProps } from './App';
import { getMovies } from '../../apiCalls/apiCalls';
import { addMovies, addLoaded, hasError } from '../../actions';

jest.mock('../../apiCalls/apiCalls');

describe("App", () => {
  const mockMoviesData =
    [{
      id: 1,
      title: "Jumanji: The Next Level",
      poster_path: 'https://some.url/poster-image',
      backdrop_path: 'https://some.url/bg-image',
      release_date: '2019-12-04',
      overview: 'In Jumanji: The Next Level, the gang is back but the game has changed.',
      average_rating: 4
    }];

  describe("App component", () => {
    let app, instance;
    const addMovies = jest.fn();
    const addLoaded = jest.fn();
    const hasError = jest.fn();

    beforeEach(() => {
      app = shallow(
        <App
          addMovies={addMovies}
          addLoaded={addLoaded}
          hasError={hasError} />
      )
      instance = app.instance()
    });

    it("should match snapshot", () => {
      expect(app).toMatchSnapshot();
    });

    it("should call fetchMoviesData method after rendering", async () => {
      const spy = jest.spyOn(instance, 'fetchMoviesData')
        .mockImplementation(() => {
          return Promise.resolve(mockMoviesData)
        });

      instance.forceUpdate()

      await instance.componentDidMount()

      expect(spy).toHaveBeenCalled();
    });

    describe("fetchMoviesData", () => {
      beforeEach(() => {
        getMovies.mockImplementation(() => {
          return Promise.resolve({movies: mockMoviesData})
        });
      });

      it("should call getMovies", async () => {
        await instance.fetchMoviesData()

        expect(getMovies).toHaveBeenCalled();
      });

      it("should call addMovies prop with movies as argument if fetch is completed correctly", async () => {

        await instance.fetchMoviesData()

        expect(addMovies).toHaveBeenCalledWith(mockMoviesData);
      });

      it("should call addLoaded with true as argument prop if fetch is completed correctly", async () => {
        await instance.fetchMoviesData()

        expect(addLoaded).toHaveBeenCalledWith(true);
      });

      it("should call addLoaded prop with false as argument if fetch is failed ", async () => {
        getMovies.mockImplementation(() => {
          return Promise.reject(Error('Failed to fetch'))
        });

        await instance.fetchMoviesData()

        expect(addLoaded).toHaveBeenCalledWith(false);
      });

      it("should call hasError prop with error as argument if fetch is failed ", async () => {
        getMovies.mockImplementation(() => {
          return Promise.reject(Error('Failed to fetch'))
        });

        await instance.fetchMoviesData()

        expect(hasError).toHaveBeenCalledWith('Failed to fetch');
      });
    });
  });

  describe('mapDispatchToProps', () => {
    let mockDispatch, mappedProps;

    beforeEach(() => {
      mockDispatch = jest.fn();
      mappedProps = mapDispatchToProps(mockDispatch);
    });

    it('calls dispatch with an addMovies action when addMovies is called', () => {
      // Setup
      const actionToDispatch = addMovies(mockMoviesData);
      // Execution
      mappedProps.addMovies(mockMoviesData);
      // Expectaion
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('calls dispatch with an addLoaded action when addLoaded is called', () => {
      // Setup
      const actionToDispatch = addLoaded(mockMoviesData);
      // Execution
      mappedProps.addLoaded(mockMoviesData);
      // Expectaion
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('calls dispatch with an hasError action when hasError is called', () => {
      // Setup
      const actionToDispatch = hasError(mockMoviesData);
      // Execution
      mappedProps.hasError(mockMoviesData);
      // Expectaion
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});
