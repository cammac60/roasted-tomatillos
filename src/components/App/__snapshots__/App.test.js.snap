// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App component should match snapshot 1`] = `
<div
  className="App"
>
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <Route
    path="/login"
    render={[Function]}
  />
</div>
`;
